### PR TITLE
Notarisation retry

### PR DESCRIPTION
**Pull request by adagys**
Add support for re-sending session messages. This is useful when talking to a distributed service, e.g. notary – if one of the nodes go down in the middle of a session, the session will be re-established with a different node (round-robin order).

I introduced a new method in `FlowLogic`: `sendAndReceiveWithRetry`, which creates a new `FlowSession` and marks it as _idempotent_. The `SessionInit` message of such session is then scheduled to be redelivered in the `NodeMessagingClient` until a `SessionData` or `SessionEnd` is received. Duplicate `SessionConfirm` responses are ignored.

### CODE REVIEW

**Comment by adagys** [comment]
Primary reviewer @shamsasari

File: core/src/main/kotlin/net/corda/core/flows/FlowLogic.kt [hunk 1]
```diff
@@ -80,6 +80,20 @@ abstract class FlowLogic<out T> {
         return stateMachine.sendAndReceive(receiveType, otherParty, payload, sessionFlow)
     }
 
+    internal inline fun <reified R : Any> sendAndReceiveWithRetry(otherParty: Party, payload: Any) = sendAndReceiveWithRetry(R::class.java, otherParty, payload)
+
+    /**
+     * Similar to [sendAndReceive] but also instructs the `payload` to be redelivered until the expected message is received.
+     * Note that this is only intended for the case when the [otherParty] is running a distributed service with an
+     * idempotent single request-response flow – e.g. a notary or certain types of oracle services. If one of more nodes
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/flows/FlowLogic.kt]: I think the "single" in "single request-response flow" is referring to the flow and not to the idea that all the flow does is a single request followed by a single response.

File: node/src/main/kotlin/net/corda/node/services/statemachine/StateMachineManager.kt [hunk 2]
```diff
@@ -301,6 +301,15 @@ class StateMachineManager(val serviceHub: ServiceHubInternal,
         val session = openSessions[message.recipientSessionId]
         if (session != null) {
             session.fiber.logger.trace { "Received $message on $session from $sender" }
+            if (session.idempotent) {
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/statemachine/StateMachineManager.kt]: I'm finding `session.idempotent` to be a bit confusing. Would `singleRequestResponse` be better?

File: node/src/main/kotlin/net/corda/node/services/statemachine/StateMachineManager.kt [hunk 3]
```diff
@@ -301,6 +301,15 @@ class StateMachineManager(val serviceHub: ServiceHubInternal,
...
+                if (message is SessionConfirm && session.state is FlowSessionState.Initiated) {
+                    session.fiber.logger.trace { "Ignoring duplicate confirmation for session ${session.ourSessionId} – session is idempotent" }
+                    return
+                }
+                if (message is SessionEnd || message is SessionData) {
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/statemachine/StateMachineManager.kt]: I think this can be just `message !is SessionConfirm`?

File: node/src/main/kotlin/net/corda/node/services/statemachine/StateMachineManager.kt [hunk 4]
```diff
@@ -518,42 +527,52 @@ class StateMachineManager(val serviceHub: ServiceHubInternal,
...
         }
     }
 
-    private fun sendSessionMessage(party: Party, message: SessionMessage, fiber: FlowStateMachineImpl<*>? = null) {
+    private fun sendSessionMessage(party: Party, message: SessionMessage, fiber: FlowStateMachineImpl<*>? = null, retryId: Long? = null) {
         val partyInfo = serviceHub.networkMapCache.getPartyInfo(party)
                 ?: throw IllegalArgumentException("Don't know about party $party")
         val address = serviceHub.networkService.getAddressOfParty(partyInfo)
         val logger = fiber?.logger ?: logger
         logger.trace { "Sending $message to party $party @ $address" }
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/statemachine/StateMachineManager.kt]: Include the retry in the trace log

File: core/src/main/kotlin/net/corda/core/flows/FlowLogic.kt [hunk 5]
```diff
@@ -80,6 +80,20 @@ abstract class FlowLogic<out T> {
...
+     * in the service cluster go down mid-session, the message will be redelivered to a different one, so there is no need
+     * to wait until the initial node comes back up to obtain a response.
+     */
+    @Suspendable
+    internal open fun <R : Any> sendAndReceiveWithRetry(receiveType: Class<R>, otherParty: Party, payload: Any): UntrustworthyData<R> {
```
Comment (mikehearn) [inline core/src/main/kotlin/net/corda/core/flows/FlowLogic.kt]: I'm not sure we do. The problem is, the name will attract usage like moths to a flame. Everyone will read it as "like the other one but more reliable" and use it. Why would you NOT want retry? Only people who read to the end of the javadoc will notice that this is talking about a rather subtle form of retry and there are in fact other forms of retry that already occur even with the normal form.

So for now I'd rather we keep the API exposed to flows smaller rather than larger.

Comment (adagys) [inline core/src/main/kotlin/net/corda/core/flows/FlowLogic.kt]: We may want to expose it in the future, but for now it can should only be used with care in very specific cases.
I was thinking the next steps would be converting things like notary, transaction resolution/data vending flows into simple request-response services (that don't use checkpointing) and we'd add an "optionally retryable service call" API to flows, which we could expose more safely.

Also I checked cordapp-template and `internal` does seem to hide these methods properly.

File: node/src/main/kotlin/net/corda/node/services/statemachine/StateMachineManager.kt (hunk 2 above)
Comment (adagys) [inline node/src/main/kotlin/net/corda/node/services/statemachine/StateMachineManager.kt]: Yes probably, but `idempotent` is also important here as we can't do SessionInit retries if the flow on the other end isn't idempotent.

File: node/src/main/kotlin/net/corda/node/services/statemachine/StateMachineManager.kt (hunk 3 above)
Comment (adagys) [inline node/src/main/kotlin/net/corda/node/services/statemachine/StateMachineManager.kt]: Yeah, but I'd rather leave it more explicit for documentation purposes.

File: node/src/main/kotlin/net/corda/node/services/statemachine/StateMachineManager.kt (hunk 2 above)
Comment (adagys) [inline node/src/main/kotlin/net/corda/node/services/statemachine/StateMachineManager.kt]: I initially had it as `retryable`, would that be better?

File: node/src/main/kotlin/net/corda/node/services/statemachine/StateMachineManager.kt (hunk 3 above)
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/statemachine/StateMachineManager.kt]: I think !is SessionConfirm is clearer. Plus your check is missing SessionRejects

File: node/src/main/kotlin/net/corda/node/services/statemachine/StateMachineManager.kt (hunk 2 above)
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/statemachine/StateMachineManager.kt]: Yes, I think so, and then we can always have some doc on the field to explain what that implies.

**Comment by adagys** [comment]
First round of comments addressed

File: core/src/main/kotlin/net/corda/core/utilities/Logging.kt [hunk 6]
```diff
@@ -20,6 +20,10 @@ inline fun org.slf4j.Logger.debug(msg: () -> String) {
     if (isDebugEnabled) debug(msg())
 }
 
+inline fun org.slf4j.Logger.warn(msg: () -> String) {
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/utilities/Logging.kt]: This would only be useful if they set their logging to just ERROR, which would not be done in prod.

**Comment by shamsasari** [comment]
It's looking good. It now needs tests.

File: core/src/main/kotlin/net/corda/core/utilities/Logging.kt (hunk 6 above)
Comment (adagys) [inline core/src/main/kotlin/net/corda/core/utilities/Logging.kt]: ah, that's true

**Comment by adagys** [comment]
I adapted one of existing tests cover the scenario where a notary node goes down in the middle of a stream of notarisation requests.
However, with our current test infrastructure I don't see a way of testing the edge case where a node goes down _after_ it confirmed a session but before it is able to send a `SessionData` response. I tested this scenario manually by inserting a slow computational task in the notary service flow and killing a node at the right point.

**Comment by exFalso** [comment]
Also, what happens if the retrying node itself is bounced? Are the retries restarted?

File: node/src/main/kotlin/net/corda/node/services/messaging/NodeMessagingClient.kt [hunk 7]
```diff
@@ -427,10 +430,48 @@ class NodeMessagingClient(override val config: NodeConfiguration,
                             "sessionID: ${message.topicSession.sessionID} uuid: ${message.uniqueMessageId}"
                 }
                 producer!!.send(mqAddress, artemisMessage)
+
+                retryId?.let {
+                    scheduledSendRetries[it] = messagingExecutor.schedule({
```
Comment (adagys) [inline node/src/main/kotlin/net/corda/node/services/messaging/NodeMessagingClient.kt]: `scheduleAtFixedRate` doesn't let you specify the max number of times to run something (or does it?). If we're e.g. talking to a single node notary and it goes down, we don't want to spam the messages indefinitely (we could work around that by only enabling retries for distributed services, but it complicates things..).

File: node/src/integration-test/kotlin/net/corda/node/services/DistributedServiceTests.kt [hunk 8]
```diff
@@ -105,27 +108,21 @@ class DistributedServiceTests : DriverBasedTest() {
...
+        val totalPayments = 20
+        for (i in 1..totalPayments) issueCash(1.POUNDS)
+        val paySelfFutures = (1..totalPayments / 2).map { paySelfAsync(1.POUNDS) }
 
         // Now kill a notary
-        with(notaries[0].process) {
-            destroy()
-            waitFor()
-        }
+        notaries[0].process.destroy()
```
Comment (exFalso) [inline node/src/integration-test/kotlin/net/corda/node/services/DistributedServiceTests.kt]: Does this trigger the retry logic? Doesn't artemis just redistribute the unacknowledged messages to other nodes?

Comment (adagys) [inline node/src/integration-test/kotlin/net/corda/node/services/DistributedServiceTests.kt]: It's not guaranteed to trigger it, but (as I mentioned it in another comment) I'm not sure how to test it end-to-end.

Comment (adagys) [inline node/src/integration-test/kotlin/net/corda/node/services/DistributedServiceTests.kt]: Added another test that does check the retries

**Comment by adagys** [comment]
Ready for another look @exFalso @shamsasari

File: node/src/integration-test/kotlin/net/corda/services/messaging/P2PMessagingTest.kt [hunk 9]
```diff
@@ -87,6 +87,100 @@ class P2PMessagingTest : NodeBasedTest() {
...
+     * Sets up the [distributedServiceNodes] to respond to [dummyTopic] requests. The first node in the service to
+     * receive a request will ignore it and all subsequent requests. This simulates the scenario where a node receives
+     * a request message, but crashes before sending back a response. The other nodes will respond to _all_ requests.
+     */
+    private fun simulateCrashingNode(distributedServiceNodes: List<Node>, dummyTopic: String, responseMessage: String): AtomicInteger {
+        val firstToReceive = AtomicBoolean(true)
+        val requestsReceived = AtomicInteger(0)
+        distributedServiceNodes.forEach {
+            val nodeName = it.info.legalIdentity.name
+            it.net.apply {
```
Comment (shamsasari) [inline node/src/integration-test/kotlin/net/corda/services/messaging/P2PMessagingTest.kt]: Can you get rid of the `apply`. I find it's a bit confusing b/c it's a large block inside a forEach. And also the fact it returns `this` is not used.

File: node/src/integration-test/kotlin/net/corda/services/messaging/P2PMessagingTest.kt [hunk 10]
```diff
@@ -87,6 +87,100 @@ class P2PMessagingTest : NodeBasedTest() {
...
+        }.getOrThrow(10.seconds)
+
+        assertThat(response).isEqualTo(responseMessage)
+    }
+
+    @Test
+    fun `distributed service request retries are persisted across client node restarts`() {
+        val distributedServiceNodes = startNotaryCluster("DistributedService", 2).getOrThrow()
+        val alice = startNode(ALICE.name, configOverrides = mapOf("messageRedeliveryDelaySeconds" to 1)).getOrThrow()
+        val serviceAddress = alice.services.networkMapCache.run {
```
Comment (shamsasari) [inline node/src/integration-test/kotlin/net/corda/services/messaging/P2PMessagingTest.kt]: What's `run` doing?

File: node/src/integration-test/kotlin/net/corda/services/messaging/P2PMessagingTest.kt [hunk 11]
```diff
@@ -87,6 +87,100 @@ class P2PMessagingTest : NodeBasedTest() {
...
+
+        // Send a single request with retry
+        with(alice.net) {
+            val request = TestRequest(sessionId, myAddress)
+            val msg = createMessage(TopicSession(dummyTopic), data = request.serialize().bytes)
+            send(msg, serviceAddress, retryId = request.sessionID)
+        }
+
+        // Wait until the first request is received
+        while (requestsReceived.get() == 0) Thread.sleep(100)
```
Comment (shamsasari) [inline node/src/integration-test/kotlin/net/corda/services/messaging/P2PMessagingTest.kt]: It think this is better achieved with a CountDownLatch.

File: node/src/main/kotlin/net/corda/node/services/messaging/NodeMessagingClient.kt [hunk 12]
```diff
@@ -429,10 +443,51 @@ class NodeMessagingClient(override val config: NodeConfiguration,
                             "sessionID: ${message.topicSession.sessionID} uuid: ${message.uniqueMessageId}"
                 }
                 producer!!.send(mqAddress, artemisMessage)
+
+                retryId?.let {
+                    database.transaction { messagesToRedeliver.computeIfAbsent(it, { Pair(message, target) }) }
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/messaging/NodeMessagingClient.kt]: This is a bit compact. Can you move the but in the tx to the next line.

File: core/src/main/kotlin/net/corda/core/messaging/Messaging.kt [hunk 13]
```diff
@@ -148,8 +154,8 @@ fun <M : Any> MessagingService.onNext(topic: String, sessionId: Long): Listenabl
 fun MessagingService.send(topic: String, sessionID: Long, payload: Any, to: MessageRecipients, uuid: UUID = UUID.randomUUID())
```
Comment (adagys) [inline core/src/main/kotlin/net/corda/core/messaging/Messaging.kt]: it's used elsewhere, and the signature is different – takes in topic & session instead of TopicSession

File: node/src/integration-test/kotlin/net/corda/services/messaging/P2PMessagingTest.kt (hunk 10 above)
Comment (adagys) [inline node/src/integration-test/kotlin/net/corda/services/messaging/P2PMessagingTest.kt]: here it uses `networkMapCache` as the receiver and returns the result of the function block.
it's to avoid `networkMapCache.getPartyInfo(networkMapCache.getAnynotary()))`

File: node/src/integration-test/kotlin/net/corda/services/messaging/P2PMessagingTest.kt (hunk 11 above)
Comment (adagys) [inline node/src/integration-test/kotlin/net/corda/services/messaging/P2PMessagingTest.kt]: not sure how to use it here, we basically first have to wait until `requestsReceived` is more than 0 to stop the node, then check that it's not more than 1 after  the node is stopped. And finally check that it's more than 2

File: node/src/integration-test/kotlin/net/corda/services/messaging/P2PMessagingTest.kt (hunk 9 above)
Comment (adagys) [inline node/src/integration-test/kotlin/net/corda/services/messaging/P2PMessagingTest.kt]: Sure

File: node/src/integration-test/kotlin/net/corda/services/messaging/P2PMessagingTest.kt (hunk 11 above)
Comment (shamsasari) [inline node/src/integration-test/kotlin/net/corda/services/messaging/P2PMessagingTest.kt]: I was thinking return the CountDownLatch along with the AtomicInteger.

Comment (adagys) [inline node/src/integration-test/kotlin/net/corda/services/messaging/P2PMessagingTest.kt]: Could do, but I feel like it's more readable this way and for the purpose of this test both approaches are functionally acceptable?

Comment (shamsasari) [inline node/src/integration-test/kotlin/net/corda/services/messaging/P2PMessagingTest.kt]: It just looks brittle with sleep polling.

File: node/src/integration-test/kotlin/net/corda/services/messaging/P2PMessagingTest.kt [hunk 14]
```diff
@@ -153,13 +155,15 @@ class P2PMessagingTest : NodeBasedTest() {
...
-    private fun simulateCrashingNode(distributedServiceNodes: List<Node>, dummyTopic: String, responseMessage: String): AtomicInteger {
+    private fun simulateCrashingNode(distributedServiceNodes: List<Node>, dummyTopic: String, responseMessage: String): Pair<CountDownLatch, AtomicInteger> {
         val firstToReceive = AtomicBoolean(true)
         val requestsReceived = AtomicInteger(0)
+        val firstRequestReceived = CountDownLatch(1)
         distributedServiceNodes.forEach {
             val nodeName = it.info.legalIdentity.name
             var ignoreRequests = false
             it.net.addMessageHandler(dummyTopic, DEFAULT_SESSION_ID) { netMessage, _ ->
+                firstRequestReceived.countDown()
```
Comment (shamsasari) [inline node/src/integration-test/kotlin/net/corda/services/messaging/P2PMessagingTest.kt]: This needs to be in the `firstToReceive` if check

Comment (adagys) [inline node/src/integration-test/kotlin/net/corda/services/messaging/P2PMessagingTest.kt]: can do, but the docs say "If the current count equals zero then nothing happens."

File: core/src/main/kotlin/net/corda/core/flows/FlowLogic.kt [hunk 15]
```diff
@@ -78,6 +78,20 @@ abstract class FlowLogic<out T> {
         return stateMachine.sendAndReceive(receiveType, otherParty, payload, sessionFlow)
     }
 
+    internal inline fun <reified R : Any> sendAndReceiveWithRetry(otherParty: Party, payload: Any) = sendAndReceiveWithRetry(R::class.java, otherParty, payload)
+
+    /**
+     * Similar to [sendAndReceive] but also instructs the `payload` to be redelivered until the expected message is received.
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/flows/FlowLogic.kt]: I think you need to add @suppress to make sure this doesn't get generated.

File: node/src/main/kotlin/net/corda/node/services/config/NodeConfiguration.kt [hunk 16]
```diff
@@ -32,6 +32,7 @@ interface NodeConfiguration : SSLConfiguration {
     val certificateSigningService: URL
     val certificateChainCheckPolicies: List<CertChainPolicyConfig>
     val verifierType: VerifierType
+    val messageRedeliveryDelaySeconds: Long
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/config/NodeConfiguration.kt]: Int is sufficient.

File: node/src/integration-test/kotlin/net/corda/services/messaging/P2PMessagingTest.kt [hunk 17]
```diff
@@ -87,6 +89,100 @@ class P2PMessagingTest : NodeBasedTest() {
...
+     */
+    private fun simulateCrashingNode(distributedServiceNodes: List<Node>, dummyTopic: String, responseMessage: String): Pair<CountDownLatch, AtomicInteger> {
+        val firstToReceive = AtomicBoolean(true)
+        val requestsReceived = AtomicInteger(0)
+        val firstRequestReceived = CountDownLatch(1)
+        distributedServiceNodes.forEach {
+            val nodeName = it.info.legalIdentity.name
+            var ignoreRequests = false
+            it.net.addMessageHandler(dummyTopic, DEFAULT_SESSION_ID) { netMessage, _ ->
+                firstRequestReceived.countDown()
```
Comment (shamsasari) [inline node/src/integration-test/kotlin/net/corda/services/messaging/P2PMessagingTest.kt]: I think this should be just after the increment, to make sure the above assert on requests received as 1 is always correct.

File: core/src/main/kotlin/net/corda/core/flows/FlowLogic.kt (hunk 15 above)
Comment (adagys) [inline core/src/main/kotlin/net/corda/core/flows/FlowLogic.kt]: I tried it out, and apparently Dokka doesn't generate docs for `internal` methods
